Userland/Libraries/LibCrypto/AEAD/ChaCha20Poly1305.cpp

```diff
@@ -0,0 +1,182 @@
+#include <LibCrypto/Authentication/Poly1305.h>
+#include <LibCrypto/Cipher/ChaCha20.h>
+
+namespace Crypto::AEAD {
+
+// https://datatracker.ietf.org/doc/html/rfc8439#section-2.6
+ErrorOr<ByteBuffer> ChaCha20Poly1305::poly1305_key()
+{
+    Crypto::Cipher::ChaCha20 cipher(m_key, m_nonce, 0);
+    cipher.generate_block();
+    auto state = cipher.block();
+    return TRY(ByteBuffer::copy(state.slice(0, 32)));
+}
+
+// https://datatracker.ietf.org/doc/html/rfc8439#section-2.8
+ErrorOr<ByteBuffer> ChaCha20Poly1305::encrypt(ReadonlyBytes aad, ReadonlyBytes input_plaintext)
+{
+    // First, a Poly1305 one-time key is generated from the 256-bit key
+    // and nonce using the procedure described in Section 2.6.
+    auto otk = TRY(poly1305_key());
+
+    // Next, the ChaCha20 encryption function is called to encrypt the
+    // plaintext, using the same key and nonce, and with the initial
+    // counter set to 1.
+    auto ciphertext_buffer = TRY(ByteBuffer::create_zeroed(input_plaintext.size()));
+    auto ciphertext = ciphertext_buffer.bytes();
+    auto chacha = Crypto::Cipher::ChaCha20(m_key, m_nonce, 1);
+    chacha.encrypt(input_plaintext, ciphertext);
+
+    // Finally, the Poly1305 function is called with the Poly1305 key
+    // calculated above, and a message constructed as a concatenation of
+    // the following:
+    auto mac_data = TRY(ByteBuffer::create_zeroed(0));
+    auto buffer_size = aad.size() + pad_to_16(aad) + ciphertext_buffer.size() + pad_to_16(ciphertext_buffer) + sizeof(u64) + sizeof(u64);
+    mac_data.ensure_capacity(buffer_size);
+
+    // The AAD
+    mac_data.append(aad);
+
+    // padding1 -- the padding is up to 15 zero bytes, and it brings
+    // the total length so far to an integral multiple of 16.  If the
+    // length of the AAD was already an integral multiple of 16 bytes,
+    // this field is zero-length.
+    for (size_t i = 0; i < pad_to_16(aad); ++i)
+        mac_data.append(0);
+
+    // The ciphertext
+    mac_data.append(ciphertext);
+
+    // padding2 -- the padding is up to 15 zero bytes, and it brings
+    // the total length so far to an integral multiple of 16.  If the
+    // length of the ciphertext was already an integral multiple of 16
+    // bytes, this field is zero-length.
+    for (size_t i = 0; i < pad_to_16(ciphertext); ++i)
+        mac_data.append(0);
+
+    u8 placeholder[8] = { 0 };
+    // The length of the additional data in octets (as a 64-bit little-endian integer).
+    mac_data.append(ReadonlyBytes { placeholder, 8 });
+    ByteReader::store(static_cast<u8*>(mac_data.end_pointer()) - sizeof(u64), AK::convert_between_host_and_little_endian(static_cast<u64>(aad.size())));
+
+    // The length of the ciphertext in octets (as a 64-bit little-endian integer).
+    mac_data.append(ReadonlyBytes { placeholder, 8 });
+    ByteReader::store(static_cast<u8*>(mac_data.end_pointer()) - sizeof(u64), AK::convert_between_host_and_little_endian(static_cast<u64>(ciphertext.size())));
```

Comment:
`mac_data.end_pointer()` should be a `u8*` already, is it not?

Reply:
Its a `void*` 🤷

Reply:
huh. well okay.